src/wallet/rpcdump.cpp

```diff
@@ -619,34 +619,58 @@ UniValue z_importkey(const UniValue& params, bool fHelp)
     if (!IsValidSpendingKey(spendingkey)) {
         throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Invalid spending key");
     }
-    // TODO: Add Sapling support. For now, ensure we can freely convert.
-    assert(boost::get<libzcash::SproutSpendingKey>(&spendingkey) != nullptr);
-    auto key = boost::get<libzcash::SproutSpendingKey>(spendingkey);
-    auto addr = key.address();
-
-    {
-        // Don't throw error in case a key is already there
-        if (pwalletMain->HaveSpendingKey(addr)) {
-            if (fIgnoreExistingKey) {
-                return NullUniValue;
+    // Sapling support
+    bool isValidKey = boost::get<libzcash::SproutSpendingKey>(&spendingkey) != nullptr || boost::get<libzcash::SaplingSpendingKey>(&spendingkey) != nullptr;
+    assert(isValidKey);
+    
+    if (boost::get<libzcash::SproutSpendingKey>(&spendingkey) != nullptr) {
+        auto key = boost::get<libzcash::SproutSpendingKey>(spendingkey);
+        auto addr = key.address();
+        {
+            // Don't throw error in case a key is already there
+            if (pwalletMain->HaveSpendingKey(addr)) {
+                if (fIgnoreExistingKey) {
+                    return NullUniValue;
+                }
+            } else {
+                pwalletMain->MarkDirty();
+                
+                if (!pwalletMain-> AddZKey(key)) {
+                    throw JSONRPCError(RPC_WALLET_ERROR, "Error adding spending key to wallet");
+                }
+                
+                pwalletMain->mapZKeyMetadata[addr].nCreateTime = 1;
             }
-        } else {
-            pwalletMain->MarkDirty();
-
-            if (!pwalletMain-> AddZKey(key))
-                throw JSONRPCError(RPC_WALLET_ERROR, "Error adding spending key to wallet");
-
-            pwalletMain->mapZKeyMetadata[addr].nCreateTime = 1;
         }
-
-        // whenever a key is imported, we need to scan the whole chain
-        pwalletMain->nTimeFirstKey = 1; // 0 would be considered 'no value'
-
-        // We want to scan for transactions and notes
-        if (fRescan) {
-            pwalletMain->ScanForWalletTransactions(chainActive[nRescanHeight], true);
+    } else {
+        auto sk = boost::get<libzcash::SaplingSpendingKey>(spendingkey);
```

Comment:
Move these sections into a `boost::static_visitor<bool>`, with the boolean indicating whether we should return early (from inside the `fIgnoreExistingKey` check). This function would become:

```cpp
...
auto keyExists = boost::apply_visitor(AddSpendingKeyToWallet(pwalletMain), spendingkey);
if (keyExists) {
    return NullUniValue;
}
// whenever a key is imported, ...
...
```